# Assignment 3-Starting point

In [23]:
# import all the needed libraries

import pandas as pd
import string
import re

import gensim         # Gensim library for word embedding 
from gensim.models import KeyedVectors

import nltk           # Natural Language Toolkit for text processing
from nltk.corpus import stopwords        # for stopwords
from nltk.stem import WordNetLemmatizer  # for lemmatization
from nltk.tokenize import TweetTokenizer # TfidfVectorizer for extraction
from sklearn.feature_extraction.text import TfidfVectorizer

In [6]:
# read the CSV file
df = pd.read_csv('master_covid19_2020_03_15-31 - master_covid19_2020_03_15-31.csv')

In [7]:
df.head()

coordinates                      created_at     hashtags  \
0         NaN  Wed Apr 01 00:09:33 +0000 2020     COVIDー19   
1         NaN  Wed Mar 18 00:24:47 +0000 2020  coronavirus   
2         NaN  Tue Mar 31 20:25:43 +0000 2020          NaN   
3         NaN  Wed Mar 18 03:32:56 +0000 2020          NaN   
4         NaN  Tue Mar 17 20:18:59 +0000 2020          NaN   

                                               media                    urls  \
0  https://twitter.com/realDonaldTrump/status/124...  http://CoronaVirus.gov   
1                                                NaN                     NaN   
2                                                NaN                     NaN   
3                                                NaN                     NaN   
4                                                NaN                     NaN   

   favorite_count            id in_reply_to_screen_name  \
0               0  1.245141e+18                     NaN   
1               0  1.240072e+18                     NaN   
2               0  1.245085e+18                     NaN   
3               0  1.240119e+18                     NaN   
4               0  1.240010e+18                     NaN   

   in_reply_to_status_id  in_reply_to_user_id  ... user_followers_count  \
0                    NaN                  NaN  ...                  922   
1                    NaN                  NaN  ...                 2287   
2                    NaN                  NaN  ...                  996   
3                    NaN                  NaN  ...                  317   
4                    NaN                  NaN  ...                 4290   

  user_friends_count user_listed_count   user_location  \
0                922                 0             NaN   
1               4996                 0             NaN   
2               1001                30  Macapá, Brasil   
3                601                 0             NaN   
4               3081                 0           706 📍   

                user_name user_screen_name.1 user_statuses_count  \
0        Cindi Puncerelli    CindiPuncerelli              172035   
1      Rosa De Ayguavives          alcanar65               40200   
2  VOTE NULO/BRANCO 🏴‍☠️♠       G_de_Almeida              292244   
3                   Ninda          nindamarz                1542   
4    11/01 ♏️🥳 SCORPI-HOE          kpetuniaa               27823   

  user_time_zone user_urls user_verified  
0            NaN       NaN         False  
1            NaN       NaN         False  
2            NaN       NaN         False  
3            NaN       NaN         False  
4            NaN       NaN         False  

[5 rows x 34 columns]

In [24]:
# display the column names
df.columns

Index(['coordinates', 'created_at', 'hashtags', 'media', 'urls',
       'favorite_count', 'id', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_user_id', 'lang', 'place',
       'possibly_sensitive', 'retweet_count', 'retweet_id',
       'retweet_screen_name', 'source', 'text', 'tweet_url', 'user_created_at',
       'user_screen_name', 'user_default_profile_image', 'user_description',
       'user_favourites_count', 'user_followers_count', 'user_friends_count',
       'user_listed_count', 'user_location', 'user_name', 'user_screen_name.1',
       'user_statuses_count', 'user_time_zone', 'user_urls', 'user_verified'],
      dtype='object')

In [5]:
# display number of rows and columns
df.shape

(24351, 34)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24351 entries, 0 to 24350
Data columns (total 34 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   coordinates                 8 non-null      object 
 1   created_at                  24351 non-null  object 
 2   hashtags                    10062 non-null  object 
 3   media                       2892 non-null   object 
 4   urls                        5402 non-null   object 
 5   favorite_count              24351 non-null  int64  
 6   id                          24351 non-null  float64
 7   in_reply_to_screen_name     1186 non-null   object 
 8   in_reply_to_status_id       1057 non-null   float64
 9   in_reply_to_user_id         1186 non-null   float64
 10  lang                        24351 non-null  object 
 11  place                       200 non-null    object 
 12  possibly_sensitive          7475 non-null   object 
 13  retweet_count               243

### 1) How many languages are available in the dataset?

In [7]:
num_languages = len(df['lang'].unique())

In [8]:
print("Number of languages in the dataset:", num_languages)

Number of languages in the dataset: 46


### 2) How many messages per each language in the data? 

In [9]:
langs_counts = df['lang'].value_counts()

In [10]:
langs_counts

en     14849
es      5538
fr      1193
pt       493
in       402
it       339
und      245
th       229
ca       213
tr       140
hi       139
de       135
tl        71
ja        62
nl        56
ar        33
lt        28
pl        23
ta        20
ru        17
ro        14
ko        11
te        11
ur        10
fi        10
zh         9
mr         9
da         6
sr         5
sv         5
ht         5
el         4
cs         4
et         4
ml         4
kn         2
bn         2
sl         2
fa         2
ne         1
si         1
gu         1
eu         1
lv         1
no         1
is         1
Name: lang, dtype: int64

There are 14849 messages in en(English), 5538 messages in es(Spanish), 1193 messages in fr(Franch), 493 messages in pt(Portuguese) and so on.

### 3) What are the top 5 most frequent hashtags?

In [11]:
hashtag_counts = df['hashtags'].value_counts()

In [12]:
hashtag_counts.head(5)

COVID19        2178
coronavirus    1195
Coronavirus     446
Covid19         228
COVIDー19        195
Name: hashtags, dtype: int64

The top 5 most frequent hashtags are "COVID19","coronavirus","Coronavirus","Covid19","COVIDー19".

### 4) Select the messages (i.e., the text field) of the langauge with the largest number of messages. Tokenize the messages in the "text" field using an appropriate tokenizer and extract the top 5 keywords using TF-IDF.

In [25]:
# select only the English tweets

df_en = df.loc[df['lang'] == 'en']['text']
df_en.head()

0    RT @realDonaldTrump: 30 DAYS TO SLOW THE SPREA...
2    RT @spectatorindex: JUST IN: 13 year old who t...
4    RT @StephMcNasty: The Coronavirus is really sh...
6    RT @CharlieDaniels: Hazel and myself send our ...
7    RT @PrakritiGaba: Last night in the ICU of a #...
Name: text, dtype: object

In [26]:
# join all the texts in the 'df_en' DataFrame into a single string
all_texts = ' '.join(df_en.tolist())

#### Pre-processing

In [17]:
# pre-processing code retrived from: https://gist.github.com/MrEliptik/b3f16179aa2f530781ef8ca9a16499af

def preprocess_text(text):
    # Remove URLs
    text = re.sub(r'http\S+', '', text)
    
    # Remove punctuation
    text = re.sub(r'[^\w\s]', '', text)
        
     # Remove word "RT" 
    text = re.sub(r'\bRT\b', '', text)
    
    # Tokenize text
    tokens = text.split()
    
    # Remove stopwords and lemmatize tokens
    stop_words = set(stopwords.words('english'))
    lemmatizer = WordNetLemmatizer()
    tokens = [lemmatizer.lemmatize(token) for token in tokens if token.lower() not in stop_words]
    
    # Rejoin tokens into a string
    text = ' '.join(tokens)
    
    return text

all_texts = preprocess_text(all_texts)

In [18]:
# stop words in English
tfidf_vectorizer = TfidfVectorizer(stop_words='english')

# compute the TF-IDF values for the texts
texts_tfidf = tfidf_vectorizer.fit_transform([all_texts])

# create a DataFrame from the TF-IDF matrix and feature names
texts_tfidf_matrix = pd.DataFrame(texts_tfidf.toarray(), columns=tfidf_vectorizer.get_feature_names_out())
print(texts_tfidf_matrix)

# convert the TF-IDF DataFrame into a dictionary
words_dict_tfidf = texts_tfidf_matrix.to_dict('index')

        000     00005       001       002       005       007        01  \
0  0.000332  0.000083  0.000083  0.000083  0.000083  0.000083  0.000166   

   012481398     01253       014  ...       𝒕𝒉𝒆      𝗕𝗝𝗣𝘀  𝗙𝗲𝗯𝗿𝘂𝗮𝗿𝘆   𝗝𝗮𝗻𝘂𝗮𝗿𝘆  \
0   0.000083  0.000083  0.000083  ...  0.000249  0.000083  0.000083  0.000083   

       𝗞𝗲𝗲𝗽     𝗠𝗮𝗿𝗰𝗵  𝗹𝗲𝗮𝗿𝗻𝗶𝗻𝗴  𝗿𝗲𝘀𝗽𝗼𝗻𝘀𝗲        𝘁𝗼        𝟭𝟳  
0  0.000083  0.000083  0.000083  0.000083  0.000083  0.000083  

[1 rows x 26929 columns]


In [20]:
# create an empty dictionary to store the words and TF-IDF score
# iterate over each text and its TF-IDF dictionary

words_tfidf = {}
for text_id, target_words in words_dict_tfidf.items():
    list_targets = [(k, v) for k, v in target_words.items()]
    list_targets_sorted = sorted(list_targets, key=lambda x: x[1], reverse=True)
    words_tfidf = list_targets_sorted[0:5]

print(words_tfidf)

[('coronavirus', 0.8796961555697431), ('covid19', 0.2736592553438731), ('case', 0.10489579347475317), ('covid', 0.10173978385318498), ('trump', 0.09650745211216404)]


In [21]:
df_top = pd.DataFrame(words_tfidf, columns=['word', 'tfidf'])
df_top

word     tfidf
0  coronavirus  0.879696
1      covid19  0.273659
2         case  0.104896
3        covid  0.101740
4        trump  0.096507

The top 5 words are: "coronavirus", "covid19", "case", "covid" and "trump".